File: Semestrul 5/PDP/lab-01/c++/OutputValidator.cpp
//
// Created by ciuis on 10/19/2025.
//

#include <iostream>
#include <fstream>
using namespace std;

class OutputValidator {
private:
    template<typename T>
    static void writeResult(T result, int rows, int cols) {
        ofstream pw(R"(C:\Users\ciuis\Desktop\FACULTATE\ANUL III\SEM I\PPD\lab-01\common\output.txt)");
        if (!pw.is_open()) {
            cerr << "FileNotFoundException" << endl;
            return;
        }

        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < cols; j++) {
                pw << result[i][j] << " ";
            }
            pw << endl;
        }
        pw.close();
    }

    template<typename T>
    static void validateImpl(T result, int rows, int cols) {
        ifstream sc(R"(C:\Users\ciuis\Desktop\FACULTATE\ANUL III\SEM I\PPD\lab-01\common\output.txt)");

        if (!sc.is_open()) {
            cerr << "FileNotFoundException" << endl;
            return;
        }

        sc.seekg(0, ios::end);
        if (sc.tellg() == 0) {
            sc.close();
            writeResult(result, rows, cols);
            return;
        }
        sc.seekg(0, ios::beg);

        bool identical = true;
        for (int i = 0; i < rows && identical; i++) {
            for (int j = 0; j < cols; j++) {
                int oldVal;
                if (!(sc >> oldVal)) {
                    identical = false;
                    break;
                }
                if (oldVal != result[i][j]) {
                    identical = false;
                    break;
                }
            }
        }
        sc.close();

        if (!identical) {
            cout << "Warning: result differs from previous run!" << endl;
        }
    }

public:
    OutputValidator() = default;

    template<typename T>
    static void validate(T result, int rows, int cols) {
        validateImpl(result, rows, cols);
    }
};

File: Semestrul 5/PDP/lab-02/java/OutputValidator.java
import java.io.File;
import java.io.FileNotFoundException;
import java.io.PrintWriter;
import java.util.Scanner;

public class OutputValidator {

    private static final File outputFile = new File("../common/output.txt");

    public OutputValidator() {
    }

    static public void validate(int[][] result) {
        try {
            if (outputFile.length() == 0) {
                writeResult(result);
                return;
            }

            Scanner sc = new Scanner(outputFile);
            boolean identical = true;

            for (int i = 0; i < result.length && identical; i++) {
                for (int j = 0; j < result[0].length; j++) {
                    if (!sc.hasNextInt()) {
                        identical = false;
                        break;
                    }
                    int oldVal = sc.nextInt();
                    if (oldVal != result[i][j]) {
                        identical = false;
                        break;
                    }
                }
            }
            sc.close();

            if (!identical) {
                System.out.println("Warning: result differs from previous run!");
            }
        } catch (FileNotFoundException e) {
            e.printStackTrace();
        }
    }

    private static void writeResult(int[][] result) {
        try (PrintWriter pw = new PrintWriter(outputFile)) {
            for (int[] row : result) {
                for (int val : row) {
                    pw.print(val + " ");
                }
                pw.println();
            }
        } catch (FileNotFoundException e) {
            e.printStackTrace();
        }
    }
}

